File: bcb6/lesson37_bcb6/Lesson37.cpp
//---------------------------------------------------------------------------

// Note: The original article for this code can be found at:
// http://www.gamedev.net/reference/programming/features/celshading

#include <windows.h>	// Header file for Windows
#include <math.h>	// Header file for the math library
#include <stdio.h>	// Header file for the standard I/O library
#include <gl\gl.h>	// Header file for the OpenGL32 library
#include <gl\glu.h>	// Header file for the GLu32 library
#include <gl\glaux.h>	// Header file for the GLaux library
#include "NeHeGL.h"	// Header file for NeHeGL

#ifndef CDS_FULLSCREEN		// CDS_FULLSCREEN is not defined by some
#define CDS_FULLSCREEN 4	// compilers. By defining it this way,
#endif				// we can avoid errors

#pragma hdrstop

//---------------------------------------------------------------------------

GL_Window* g_window;
Keys* g_keys;

// User defined structures
typedef struct tagMATRIX	// A structure to hold an OpenGL matrix
{
	float Data[16];		// We use [16] due to OpenGL's matrix format
}
MATRIX;

typedef struct tagVECTOR	// A structure to hold a single vector
{
	float X, Y, Z;		// The components of the vector
}
VECTOR;

typedef struct tagVERTEX	// A structure to hold a single vertex
{
	VECTOR Nor;		// Vertex normal
	VECTOR Pos;	        // Vertex position
}
VERTEX;

typedef struct tagPOLYGON	// A structure to hold a single polygon
{
	VERTEX Verts[3];	// Array of 3 VERTEX structures
}
POLYGON;

// User defined variables
bool outlineDraw  = true;	// Flag to draw the outline
bool outlineSmooth = false;	// Flag to anti-alias the lines
float outlineColor[3] = { 0.0f, 0.0f, 0.0f };	// Color of the lines
float outlineWidth = 3.0f;	// Width of the lines

VECTOR lightAngle;		// The direction of the light
bool lightRotate = false;	// Flag to see if we rotate the light

float modelAngle = 0.0f;	// Y-axis angle of the model
bool modelRotate = false;	// Flag to rotate the model

POLYGON *polyData = NULL;	// Polygon data
int polyNum = 0;		// Number of polygons

GLuint shaderTexture[1];        // Storage for one texture

// File functions
BOOL ReadMesh ()		// Reads the contents of the "model.txt" file
{
	FILE *In = fopen ("Data\\model.txt", "rb");	// Open the file

	if (!In)
		return FALSE;				// Return FALSE if file not opened

	fread (&polyNum, sizeof (int), 1, In);		// Read the header (i.e. number of polygons)

	polyData = new POLYGON [polyNum];		// Allocate the memory

	fread (&polyData[0], sizeof (POLYGON) * polyNum, 1, In);	// Read in all polygon data

	fclose (In);	        // Close the file

	return true;		// It worked
}

// Math Functions
inline float DotProduct (VECTOR &V1, VECTOR &V2)	// Calculate the angle between the 2 vectors
{
	return V1.X * V2.X + V1.Y * V2.Y + V1.Z * V2.Z;	// Return the angle
}

inline float Magnitude (VECTOR &V)	// Calculate the length of the vector
{
	return sqrt (V.X * V.X + V.Y * V.Y + V.Z * V.Z);	// Return the length of the vector
}

void Normalize (VECTOR &V)		// Creates a vector with a unit length of 1
{
	float M = Magnitude (V);	// Calculate the length of the vector

	if (M != 0.0f)			// Make sure we don't divide by 0
	{
		V.X /= M;		// Normalize the 3 components
		V.Y /= M;
		V.Z /= M;
	}
}

void RotateVector (MATRIX &M, VECTOR &V, VECTOR &D)	        // Rotate a vector using the supplied matrix
{
	D.X = (M.Data[0] * V.X) + (M.Data[4] * V.Y) + (M.Data[8]  * V.Z);	// Rotate around the X axis
	D.Y = (M.Data[1] * V.X) + (M.Data[5] * V.Y) + (M.Data[9]  * V.Z);	// Rotate around the Y axis
	D.Z = (M.Data[2] * V.X) + (M.Data[6] * V.Y) + (M.Data[10] * V.Z);	// Rotate around the Z axis
}

// Engine Functions
BOOL Initialize (GL_Window* window, Keys* keys)	// Any GL init code & user initialiazation goes here
{
	int i;				// Looping variable
	char Line[255];			// Storage for 255 characters
	float shaderData[32][3];	// Storate for the 96 shader values

	FILE *In = NULL;		// File pointer

	g_window = window;
	g_keys = keys;

	// Start Of User Initialization
	glHint (GL_PERSPECTIVE_CORRECTION_HINT, GL_NICEST);	// Realy nice perspective calculations

	glClearColor (0.7f, 0.7f, 0.7f, 0.0f);			// Light grey background
	glClearDepth (1.0f);			// Depth buffer setup

	glEnable (GL_DEPTH_TEST);		// Enable depth testing
	glDepthFunc (GL_LESS);			// The type of depth test to do

	glShadeModel (GL_SMOOTH);		// Enables smooth color shading
	glDisable (GL_LINE_SMOOTH);		// Initially disable line smoothing

	glEnable (GL_CULL_FACE);		// Enable OpenGL face culling

	glDisable (GL_LIGHTING);		// Disable OpenGL lighting

	In = fopen ("Data\\shader.txt", "r");	// Open the shader file

	if (In)					// Check to see if the file opened
	{
		for (i = 0; i < 32; i++)	// Loop though the 32 greyscale values
		{
			if (feof (In))		// Check for the end of the file
				break;

			fgets (Line, 255, In);	// Get the current line

			shaderData[i][0] = shaderData[i][1] = shaderData[i][2] = float(atof (Line)); // Copy over the value
		}

		fclose (In);		        // Close the file
	}

	else
		return false;		        // It went horribly horribly wrong

	glGenTextures (1, &shaderTexture[0]);	// Get a free texture ID

	glBindTexture (GL_TEXTURE_1D, shaderTexture[0]);	// Bind this texture. From now on it will be 1D

	// For crying out loud don't let OpenGL use bi/trilinear filtering!
	glTexParameteri (GL_TEXTURE_1D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
	glTexParameteri (GL_TEXTURE_1D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);

	glTexImage1D (GL_TEXTURE_1D, 0, GL_RGB, 32, 0, GL_RGB , GL_FLOAT, shaderData);	// Upload

	lightAngle.X = 0.0f;		// Set the X direction
	lightAngle.Y = 0.0f;		// Set the Y direction
	lightAngle.Z = 1.0f;		// Set the Z direction

	Normalize (lightAngle);		// Normalize the light direction

	return ReadMesh ();		// Return the value of readmesh
}

void Deinitialize (void)		// Any user deinitialization goes here
{
	glDeleteTextures (1, &shaderTexture[0]);	// Delete the shader texture

	delete [] polyData;				// Delete the polygon data
}

void Update (DWORD milliseconds)			// Perform motion updates here
{
	if (g_keys->keyDown [VK_ESCAPE] == TRUE)	// Is ESC being pressed?
	{
		TerminateApplication (g_window);	// Terminate the program
	}

	if (g_keys->keyDown [VK_F1] == TRUE)		// Is F1 being pressed?
	{
		ToggleFullscreen (g_window);		// Toggle fullscreen mode
	}

	if (g_keys->keyDown [' '] == TRUE)		// Is the space bar being pressed?
	{
		modelRotate = !modelRotate;		// Toggle model rotation on/off

		g_keys->keyDown [' '] = FALSE;
	}

	if (g_keys->keyDown ['1'] == TRUE)		// Is the number 1 being pressed?
	{
		outlineDraw = !outlineDraw;		// Toggle outline drawing on/off

		g_keys->keyDown ['1'] = FALSE;
	}

	if (g_keys->keyDown ['2'] == TRUE)		// Is the number 2 being pressed?
	{
		outlineSmooth = !outlineSmooth;		// Toggle anti-aliasing on/off

		g_keys->keyDown ['2'] = FALSE;
	}

	if (g_keys->keyDown [VK_UP] == TRUE)		// Is the up arrow being pressed?
	{
		outlineWidth++;				// Increase line width

		g_keys->keyDown [VK_UP] = FALSE;
	}

	if (g_keys->keyDown [VK_DOWN] == TRUE)		// Is the down arrow being pressed?
	{
		outlineWidth--;				// Decrease line width

		g_keys->keyDown [VK_DOWN] = FALSE;
	}

	if (modelRotate)				// Check to see if rotation is enabled
		modelAngle += (float) (milliseconds) / 10.0f;	// Update angle based on the clock
}

void Draw (void)
{
	int i, j;			// Looping variables

	float TmpShade;			// Temporary shader value

	MATRIX TmpMatrix;		// Temporary MATRIX structure
	VECTOR TmpVector, TmpNormal;	// Temporary VECTOR structures
	
	glClear (GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);	// Clear the buffers
	glLoadIdentity ();					// Reset the matrix

	if (outlineSmooth)					// Check to see if we want anti-aliased lines
	{
		glHint (GL_LINE_SMOOTH_HINT, GL_NICEST);	// Use the good calculations
		glEnable (GL_LINE_SMOOTH);			// Enable anti-aliasing
	}

	else							// We don't want smooth lines
		glDisable (GL_LINE_SMOOTH);			// Disable anti-aliasing

	glTranslatef (0.0f, 0.0f, -2.0f);			// Move 2 units away from the screen
	glRotatef (modelAngle, 0.0f, 1.0f, 0.0f);		// Rotate the model on it's Y-axis

	glGetFloatv (GL_MODELVIEW_MATRIX, TmpMatrix.Data);	// Get the generated matrix

	// Cel-Shading Code //
	glEnable (GL_TEXTURE_1D);				// Enable 1D texturing
	glBindTexture (GL_TEXTURE_1D, shaderTexture[0]);	// Bind our texture

	glColor3f (1.0f, 1.0f, 1.0f);			// Set the color of the model

	glBegin (GL_TRIANGLES);				// Tell OpenGL that we're drawing triangles

		for (i = 0; i < polyNum; i++)		// Loop through each polygon
		{
			for (j = 0; j < 3; j++)		// Loop through each vertex
			{
				TmpNormal.X = polyData[i].Verts[j].Nor.X;	// Fill up the TmpNormal structure with
				TmpNormal.Y = polyData[i].Verts[j].Nor.Y;	// the current vertices' normal values
				TmpNormal.Z = polyData[i].Verts[j].Nor.Z;

				RotateVector (TmpMatrix, TmpNormal, TmpVector);	// Rotate this by the matrix

				Normalize (TmpVector);				// Normalize the new normal

				TmpShade = DotProduct (TmpVector, lightAngle);	// Calculate the shade value

				if (TmpShade < 0.0f)
					TmpShade = 0.0f;			// Clamp the value to 0 if negative

				glTexCoord1f (TmpShade);			// Set the texture co-ordinate as the shade value
				glVertex3fv (&polyData[i].Verts[j].Pos.X);	// Send the vertex position
		    }
		}

    glEnd ();							// Tell OpenGL to finish drawing

	glDisable (GL_TEXTURE_1D);				// Disable 1D textures

	// Outline Code //
	if (outlineDraw)					// Check to see if we want to draw the outline
	{
		glEnable (GL_BLEND);					// Enable blending
		glBlendFunc(GL_SRC_ALPHA,GL_ONE_MINUS_SRC_ALPHA);	// Set the blend mode

		glPolygonMode (GL_BACK, GL_LINE);	        // Draw backfacing polygons as wireframes
		glLineWidth (outlineWidth);			// Set the line width

		glCullFace (GL_FRONT);				// Don't draw any front-facing polygons

		glDepthFunc (GL_LEQUAL);			// Change the depth mode

		glColor3fv (&outlineColor[0]);			// Set the outline color

		glBegin (GL_TRIANGLES);				// Tell OpenGL what we want to draw

			for (i = 0; i < polyNum; i++)		// Loop through each polygon
			{
				for (j = 0; j < 3; j++)		// Loop through each vertex
				{
					glVertex3fv (&polyData[i].Verts[j].Pos.X);	// Send the vertex position
				}
			}

		glEnd ();				// Tell OpenGL we've finished

		glDepthFunc (GL_LESS);			// Reset the depth-testing mode

		glCullFace (GL_BACK);			// Reset the face to be culled

		glPolygonMode (GL_BACK, GL_FILL);	// Reset back-facing polygon drawing mode

		glDisable (GL_BLEND);			// Disable blending
	}
}

File: jogl/lesson37/ztv807/Lesson37.java
                  /*--.          .-"-.
                 /   o_O        / O o \
                 \_  (__\       \_ v _/
                 //   \\        //   \\
                ((     ))      ((     ))
 ¤¤¤¤¤¤¤¤¤¤¤¤¤¤--""---""--¤¤¤¤--""---""--¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤
 ¤                 |||            |||                             ¤
 ¤                  |              |                              ¤
 ¤                                                                ¤
 ¤ Programmer:Abdul Bezrati                                       ¤
 ¤ Program   :Nehe's 37th lesson port to JOGL                     ¤
 ¤ Comments  :None                                                ¤
 ¤    _______                                                     ¤
 ¤  /` _____ `\;,    abezrati@hotmail.com                         ¤
 ¤ (__(^===^)__)';,                                 ___           ¤
 ¤   /  :::  \   ,;                               /^   ^\         ¤
 ¤  |   :::   | ,;'                              ( Ö   Ö )        ¤
 ¤¤¤'._______.'`¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤¤ --°oOo--(_)--oOo°--¤¤*/

import net.java.games.jogl.*;
import net.java.games.jogl.util.GLUT;
import java.util.StringTokenizer;
import java.awt.event.*;
import javax.swing.*;
import java.net.URL;
import java.awt.*;
import java.io.*;

public class Lesson37 implements KeyListener
{
  initRenderer      renderer;
  GLCanvas          canvas;
  Animator          loop;
  JFrame            frame;
  GLU               glu;
  GL                gl;

  boolean           outlineSmooth,                           // Flag To Anti-Alias The Lines ( NEW )
                    outlineDraw = true,                      // Flag To Draw The Outline ( NEW )
                    lightRotate = false,                     // Flag To See If We Rotate The Light ( NEW )
                    modelRotate = false,                     // Flag To Rotate The Model ( NEW )
                    keys[]      = new boolean[256],          // Array Used For The Keyboard Routine
                    light;                                   // Lighting ON/OFF
  // User Defined Variables
  float             outlineColor[] = { 0.0f, 0.0f, 0.0f },   // Color Of The Lines ( NEW )
                    outlineWidth   = 3f,                     // Width Of The Lines ( NEW )
                    modelAngle     = 0f;                     // Y-Axis Angle Of The Model ( NEW )

  POLYGON           polyData[];                              // Polygon Data ( NEW )
  VECTOR            lightAngle = new VECTOR();               // The Direction Of The Light ( NEW )

  int               shaderTexture[] = new int[1],            // Storage For One Texture ( NEW )
                    screenWidth,
                    screenHeight,
                    canvasHeight,
                    canvasWidth,
                    xLocation,
                    yLocation,
                    polyNum         = 0;                     // Number Of Polygons ( NEW )

  // User Defined Structures
  class MATRIX{ float Data[] = new float[16];}               // A Structure To Hold An OpenGL Matrix ( NEW )
                                                             // We Use [16] Due To OpenGL's Matrix Format ( NEW )
  class VECTOR{float X, Y, Z;}                               // A Structure To Hold A Single Vector ( NEW )
  class VERTEX{                                              // A Structure To Hold A Single Vertex ( NEW )
                VECTOR Nor = new VECTOR(),                   // Vertex Normal ( NEW )
                       Pos = new VECTOR();                   // Vertex Position ( NEW )
  }
  class POLYGON{                                             // A Structure To Hold A Single Polygon ( NEW )
     VERTEX Verts[] = new VERTEX[3];                         // Array Of 3 VERTEX Structures ( NEW )
     POLYGON(){
       for(int i =0; i<3; i++)
         Verts[i] = new VERTEX();
     }
  }

  // Math Functions
  float DotProduct(VECTOR V1, VECTOR V2){                       // Calculate The Angle Between The 2 Vectors ( NEW )
    return V1.X * V2.X + V1.Y * V2.Y + V1.Z * V2.Z;             // Return The Angle ( NEW )
  }

  float Magnitude(VECTOR V){                                    // Calculate The Length Of The Vector ( NEW )
    return (float)Math.sqrt(V.X * V.X + V.Y * V.Y + V.Z * V.Z); // Return The Length Of The Vector ( NEW )
  }

  void Normalize(VECTOR V){                                     // Creates A Vector With A Unit Length Of 1 ( NEW )

    float M = Magnitude (V);                                    // Calculate The Length Of The Vector  ( NEW )

    if(M != 0.0f){                                              // Make Sure We Don't Divide By 0  ( NEW )
      V.X /= M;                                                 // Normalize The 3 Components  ( NEW )
      V.Y /= M;
      V.Z /= M;
    }
  }

  void RotateVector(MATRIX M, VECTOR V, VECTOR D){              // Rotate A Vector Using The Supplied Matrix ( NEW )

    D.X = (M.Data[0] * V.X) + (M.Data[4] * V.Y) + (M.Data[8]  * V.Z);  // Rotate Around The X Axis ( NEW )
    D.Y = (M.Data[1] * V.X) + (M.Data[5] * V.Y) + (M.Data[9]  * V.Z);  // Rotate Around The Y Axis ( NEW )
    D.Z = (M.Data[2] * V.X) + (M.Data[6] * V.Y) + (M.Data[10] * V.Z);  // Rotate Around The Z Axis ( NEW )
  }

  public static void main(String []args) {
    Lesson37 demo = new Lesson37();
  }

  Lesson37(){
    int fullScreen = JOptionPane.showConfirmDialog(        null, "Would you like to run in fullscreen mode?",
                                                   "Fullscreen",  JOptionPane.YES_NO_OPTION);
    if(fullScreen!=0)
      JFrame.setDefaultLookAndFeelDecorated(true);

    frame          = new JFrame("NeHe's Cel-Shading Tutorial");
    screenWidth    = Toolkit.getDefaultToolkit().getScreenSize().width;
    screenHeight   = Toolkit.getDefaultToolkit().getScreenSize().height;

    switch(fullScreen){
      case 0:
        frame.setUndecorated(true);
      break;
      default:
        canvasWidth  = 640;
        canvasHeight = 480;
        xLocation    = (screenWidth  - canvasWidth )>>1;
        yLocation    = (screenHeight - canvasHeight)>>1;
        frame.setLocation(xLocation,yLocation);
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setIconImage(new ImageIcon("Data/icon.png").getImage());
    }
    canvas  = GLDrawableFactory.getFactory().createGLCanvas(new GLCapabilities());
    canvas.setSize(new Dimension(canvasWidth,canvasHeight));
    canvas.addGLEventListener((renderer = new initRenderer()));
    canvas.requestFocus();
    canvas.addKeyListener(this);

    frame.addKeyListener(this);
    frame.addWindowListener(new shutDownWindow());
    frame.getContentPane().add(canvas,BorderLayout.CENTER);

    if(fullScreen==0){
      GraphicsEnvironment.getLocalGraphicsEnvironment().
      getDefaultScreenDevice().setFullScreenWindow(frame);
      GraphicsEnvironment.getLocalGraphicsEnvironment().
      getDefaultScreenDevice().setDisplayMode((new DisplayMode(640, 480, 32,
                                               DisplayMode.REFRESH_RATE_UNKNOWN)));
    }
    else
      frame.pack();
    frame.setVisible(true);
  }

  public class initRenderer
               implements GLEventListener
  {
    public void init(GLDrawable drawable){

      gl         = drawable.getGL();
      glu        = drawable.getGLU();
      int   i,                                                    // Looping Variable ( NEW )
            info;
      float shaderData[]  = new float[96];                        // Storate For The 96 Shader Values ( NEW )

      // Start Of User Initialization
      gl.glHint(gl.GL_PERSPECTIVE_CORRECTION_HINT, gl.GL_NICEST); // Realy Nice perspective calculations
      gl.glClearColor(0.7f, 0.7f, 0.7f, 0.0f);                    // Light Grey Background
      gl.glClearDepth(1.0f);                                      // Depth Buffer Setup

      gl.glEnable(gl.GL_DEPTH_TEST);                              // Enable Depth Testing
      gl.glDepthFunc(gl.GL_LESS);                                 // The Type Of Depth Test To Do

      gl.glShadeModel(gl.GL_SMOOTH);                              // Enables Smooth Color Shading ( NEW )
      gl.glDisable(gl.GL_LINE_SMOOTH);                            // Initially Disable Line Smoothing ( NEW )

      gl.glEnable(gl.GL_CULL_FACE);                               // Enable OpenGL Face Culling ( NEW )

      gl.glDisable(gl.GL_LIGHTING);                               // Disable OpenGL Lighting ( NEW )

      FileReader   shader;
      StringBuffer readShaderData = new StringBuffer();

      try{
        shader = new FileReader("Data\\shader.txt");              // Open The Shader File ( NEW )
        while ((info = shader.read()) != -1)
        readShaderData.append((char)info);
        shader.close();
      }
      catch(IOException e){
        System.out.println("Couldn't open shader.txt");
        System.exit(0);
      }

      StringTokenizer tokenizer = new StringTokenizer(readShaderData.toString());

      for(i = 0; i < 96; i+=3)                                      // Loop Though The 32 Greyscale Values ( NEW )
        shaderData[i  ] =
        shaderData[i+1] =
        shaderData[i+2] = Float.parseFloat(tokenizer.nextToken());  // Copy Over The Value ( NEW )

      gl.glGenTextures(1, shaderTexture);                           // Get A Free Texture ID ( NEW )
      gl.glBindTexture(gl.GL_TEXTURE_1D, shaderTexture[0]);         // Bind This Texture. From Now On It Will Be 1D ( NEW )

      // For Crying Out Loud Don't Let OpenGL Use Bi/Trilinear Filtering! ( NEW )
      gl.glTexParameteri(gl.GL_TEXTURE_1D,gl.GL_TEXTURE_MAG_FILTER,gl.GL_NEAREST);
      gl.glTexParameteri(gl.GL_TEXTURE_1D,gl.GL_TEXTURE_MIN_FILTER,gl.GL_NEAREST);

      gl.glTexImage1D(gl.GL_TEXTURE_1D, 0,gl.GL_RGB, 32, 0,gl.GL_RGB ,gl.GL_FLOAT, shaderData); // Upload ( NEW )

      lightAngle.X = 0.0f;                                          // Set The X Direction ( NEW )
      lightAngle.Y = 0.0f;                                          // Set The Y Direction ( NEW )
      lightAngle.Z = 1.0f;                                          // Set The Z Direction ( NEW )

      Normalize(lightAngle);                                        // Normalize The Light Direction ( NEW )
      ReadMesh ();                                                  // Return The Value Of ReadMesh ( NEW )

      loop = new Animator(drawable);
      loop.start();
    }

    public void display(GLDrawable drawable){
      // Clear Color Buffer, Depth Buffer
      gl.glClear(GL.GL_COLOR_BUFFER_BIT | GL.GL_DEPTH_BUFFER_BIT );
      int   i, j;                                                 // Looping Variables ( NEW )
      float TmpShade;                                             // Temporary Shader Value ( NEW )

      MATRIX TmpMatrix = new MATRIX();                            // Temporary MATRIX Structure ( NEW )
      VECTOR TmpVector = new VECTOR(),
             TmpNormal = new VECTOR();                            // Temporary VECTOR Structures ( NEW )

      gl.glLoadIdentity();                                        // Reset The Matrix

      if(outlineSmooth){                                          // Check To See If We Want Anti-Aliased Lines ( NEW )
        gl.glHint(gl.GL_LINE_SMOOTH_HINT,gl.GL_NICEST);           // Use The Good Calculations ( NEW )
        gl.glEnable(gl.GL_LINE_SMOOTH);                           // Enable Anti-Aliasing ( NEW )
      }
      else                                                        // We Don't Want Smooth Lines ( NEW )
        gl.glDisable(gl.GL_LINE_SMOOTH);                          // Disable Anti-Aliasing ( NEW )

      gl.glTranslatef (0.0f, 0.0f, -2.0f);                        // Move 2 Units Away From The Screen ( NEW )
      gl.glRotatef(modelAngle, 0.0f, 1.0f, 0.0f);                 // Rotate The Model On It's Y-Axis ( NEW )

      gl.glGetFloatv(gl.GL_MODELVIEW_MATRIX, TmpMatrix.Data);     // Get The Generated Matrix ( NEW )

      // Cel-Shading Code //
      gl.glEnable(gl.GL_TEXTURE_1D);                              // Enable 1D Texturing ( NEW )
      gl.glBindTexture(gl.GL_TEXTURE_1D, shaderTexture[0]);       // Bind Our Texture ( NEW )
      gl.glColor3f(1.0f, 1.0f, 1.0f);                             // Set The Color Of The Model ( NEW )

      gl.glBegin(gl.GL_TRIANGLES);                                // Tell OpenGL That We're Drawing Triangles

      for(i = 0; i < polyNum; i++)                                // Loop Through Each Polygon ( NEW )
        for(j = 0; j < 3; j++){                                   // Loop Through Each Vertex ( NEW )

          TmpNormal.X = polyData[i].Verts[j].Nor.X;               // Fill Up The TmpNormal Structure With
          TmpNormal.Y = polyData[i].Verts[j].Nor.Y;               // The Current Vertices' Normal Values ( NEW )
          TmpNormal.Z = polyData[i].Verts[j].Nor.Z;
          RotateVector (TmpMatrix, TmpNormal, TmpVector);         // Rotate This By The Matrix ( NEW )
          Normalize(TmpVector);                                   // Normalize The New Normal ( NEW )

          TmpShade = DotProduct(TmpVector, lightAngle);           // Calculate The Shade Value ( NEW )
          if(TmpShade < 0.0f)
            TmpShade = 0.0f;                                      // Clamp The Value to 0 If Negative ( NEW )

          gl.glTexCoord1f(TmpShade);                              // Set The Texture Co-ordinate As The Shade Value ( NEW )
          gl.glVertex3f(polyData[i].Verts[j].Pos.X,
                        polyData[i].Verts[j].Pos.Y,
                        polyData[i].Verts[j].Pos.Z);              // Send The Vertex Position ( NEW )
        }

      gl.glEnd ();                                                // Tell OpenGL To Finish Drawing
      gl.glDisable(gl.GL_TEXTURE_1D);                             // Disable 1D Textures ( NEW )

      // Outline Code //
      if(outlineDraw){                                            // Check To See If We Want To Draw The Outline ( NEW )
        gl.glEnable(gl.GL_BLEND);                                 // Enable Blending ( NEW )
        gl.glBlendFunc(gl.GL_SRC_ALPHA,gl.GL_ONE_MINUS_SRC_ALPHA);// Set The Blend Mode ( NEW )

        gl.glPolygonMode(gl.GL_BACK,gl.GL_LINE);                  // Draw Backfacing Polygons As Wireframes ( NEW )
        gl.glLineWidth(outlineWidth);                             // Set The Line Width ( NEW )
        gl.glCullFace(gl.GL_FRONT);                               // Don't Draw Any Front-Facing Polygons ( NEW )

        gl.glDepthFunc(gl.GL_LEQUAL);                             // Change The Depth Mode ( NEW )
        gl.glColor3fv(outlineColor);                              // Set The Outline Color ( NEW )

        gl.glBegin(gl.GL_TRIANGLES);                              // Tell OpenGL What We Want To Draw

        for(i = 0; i < polyNum; i++)                              // Loop Through Each Polygon ( NEW )
          for(j = 0; j < 3; j++)                                  // Loop Through Each Vertex ( NEW )
            gl.glVertex3f(polyData[i].Verts[j].Pos.X,
                          polyData[i].Verts[j].Pos.Y,
                          polyData[i].Verts[j].Pos.Z);            // Send The Vertex Position ( NEW )

        gl.glEnd();                                               // Tell OpenGL We've Finished
        gl.glDepthFunc(gl.GL_LESS);                               // Reset The Depth-Testing Mode ( NEW )
        gl.glCullFace(gl.GL_BACK);                                // Reset The Face To Be Culled ( NEW )
        gl.glPolygonMode(gl.GL_BACK,gl.GL_FILL);                  // Reset Back-Facing Polygon Drawing Mode ( NEW )
        gl.glDisable(gl.GL_BLEND);                                // Disable Blending ( NEW )
      }
      if(modelRotate)                                              // Check To See If Rotation Is Enabled ( NEW )
        modelAngle += .2f;                                      // Update Angle Based On The Clock
    }

    public void reshape(GLDrawable drawable,
                        int xstart,int ystart,
                        int width, int height){

      height = (height == 0) ? 1 : height;

      gl.glViewport(0,0,width,height);
      gl.glMatrixMode(gl.GL_PROJECTION);
      gl.glLoadIdentity();

      glu.gluPerspective(45,(float)width/height,1,1000);
      gl.glMatrixMode(gl.GL_MODELVIEW);
      gl.glLoadIdentity();
    }

    public void displayChanged(GLDrawable drawable,
                               boolean modeChanged,
                               boolean deviceChanged){}
  }

  void ReadMesh(){

    URL fileSource = null;
    try{
      fileSource = new URL("file", "localhost","Data\\model.txt");
    }
    catch(java.net.MalformedURLException e){
      System.out.println("Couldn't locate model, exiting");
      System.exit(0);
    }

    InputStream in = null;
    try {
      in = fileSource.openStream();
    }
    catch(IOException e){
      System.out.println("Couldn't load model, exiting");
      System.exit(0);
    }

    polyNum  = byteToInt(readNextFourBytes(in));
    polyData = new POLYGON[polyNum];

    for(int i=0; i<polyData.length;i++){
      polyData[i] = new POLYGON();
      for(int j =0; j<3; j++){
        polyData[i].Verts[j].Nor.X = byteToFloat(readNextFourBytes(in));
        polyData[i].Verts[j].Nor.Y = byteToFloat(readNextFourBytes(in));
        polyData[i].Verts[j].Nor.Z = byteToFloat(readNextFourBytes(in));

        polyData[i].Verts[j].Pos.X = byteToFloat(readNextFourBytes(in));
        polyData[i].Verts[j].Pos.Y = byteToFloat(readNextFourBytes(in));
        polyData[i].Verts[j].Pos.Z = byteToFloat(readNextFourBytes(in));
      }
    }
  }

  byte[] readNextFourBytes(InputStream in){
    byte[] bytes = new byte[4];

    try{
      for(int i = 0; i<4; i++)
        bytes[i] = (byte)in.read();
    }
    catch(Exception e){
      System.out.println("Error reading file, exiting");
      System.exit(0);
    }
    return bytes;
  }

  int byteToInt(byte[] array) {
    int value = 0;
    for(int i = 0; i < 4; i++) {
      int b = array[i];
      b &= 0xff;
      value |= (b << (i*8));
    }
   return value;
  }

  float byteToFloat(byte[] array){
    int value = 0;
    for(int i = 3; i >= 0; i--) {
      int b = array[i];
      b &= 0xff;
      value |= (b << (i * 8));
    }
    return Float.intBitsToFloat(value);
  }

  public void keyReleased(KeyEvent evt){

    keys[evt.getKeyCode()] = false;
  }

  public void keyPressed (KeyEvent evt){

    keys[evt.getKeyCode()] = true;

    if(keys[KeyEvent.VK_ESCAPE]){
      loop.stop();
      System.exit(0);
    }

    if(keys[KeyEvent.VK_SPACE])                         // Is the Space Bar Being Pressed? ( NEW )
      modelRotate = !modelRotate;                       // Toggle Model Rotation On/Off ( NEW )

    if(keys[KeyEvent.VK_1])                             // Is The Number 1 Being Pressed? ( NEW )
      outlineDraw = !outlineDraw;                       // Toggle Outline Drawing On/Off ( NEW )

    if(keys[KeyEvent.VK_2])                             // Is The Number 2 Being Pressed? ( NEW )
      outlineSmooth = !outlineSmooth;                   // Toggle Anti-Aliasing On/Off ( NEW )

    if(keys[KeyEvent.VK_UP])                            // Is The Up Arrow Being Pressed? ( NEW )
      outlineWidth++;                                   // Increase Line Width ( NEW )

    if(keys[KeyEvent.VK_DOWN])                          // Is The Down Arrow Being Pressed? ( NEW )
      outlineWidth--;                                   // Decrease Line Width ( NEW )
 }

  public void keyTyped   (KeyEvent evt){}
  public class shutDownWindow extends WindowAdapter {
    public void windowClosing(WindowEvent e) {
      loop.stop();
    }
  }
}
